modules/spring-config/src/main/java/org/mule/runtime/config/internal/BeanWrapper.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.config.internal;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * BeanWrapper helps the comparison/equality check among non-proxy vs. proxy or proxy vs. proxy when building a dependency graph.
+ */
+public class BeanWrapper {
+
+  private final Object wrappedObject;
+  private final String name;
+
+  public BeanWrapper(String name, Object o) {
+    requireNonNull(o, "bean must not be null");
+    this.name = name;
+    this.wrappedObject = o;
+  }
+
+  public Object getWrappedObject() {
+    return wrappedObject;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (wrappedObject.equals(obj)) {
+      return true;
+    } else if (obj instanceof BeanWrapper) {
+      return this.wrappedObject.hashCode() == ((BeanWrapper) obj).wrappedObject.hashCode();
```

Comment:
this.wrappedObject.equals(((BeanWrapper) obj).wrappedObject) was not able to tell if two proxy objects are same when they wrap the same object. Using the hashCodes kind of solved this problem.

Reply:
This doesn't really make sense, if the wrapped objects are the same then `this.wrappedObject.equals(((BeanWrapper) obj).wrappedObject)` should return true, because you would be comparing the wrapped objects, not the wrappers.
If using the `hashCode` changes anything then it means the issue is coming from the fact that two distinct objects may have the same `hashCode`, which is perfectly valid, but I don't think that we should make our equality check be dependant on that (because of the possible collisions that Eze mentioned).
Are you absolutely sure that you have already tried Eze's suggestion and not something else?

Reply:
In that case, I need help to move away from the hashcode comparison. If I use `this.wrappedObject.equals(((BeanWrapper) obj).wrappedObject)` as suggested, I have seen some actual cases where we cannot detect the same objects if they are in different forms. There are some examples in `BeanWrapperTestCase.java`, such as ```
sameProxyInWrapperComparisonTest
sameProxyInTwoDifferentWrappersComparisonTest
objectInProxyAndWrapperComparisonTest
```

Reply:
I think we need to understand what happens in the real case, because in the tests you are using a `FakeProxy` which has its own `equals` implementation which is delegating the equals against the real object directly. So, two different proxies that delegate to the same real object will have the same `hashCode` but will not compare equal. Not only that, I think your `FakeProxy` will not compare equal to itself (`x.equals(x)`), which violates the reflexivity property of the equals contract.